File: scripts/cleanup-dod.py
#!/usr/bin/env python3
"""
ICE — Replace inline DoD with reference to ICE_ENGINEERING.md
Uses urllib (not node fetch) to avoid DNS resolution issues.
"""
import json, urllib.request, time, sys

API_KEY = "lin_api_FEzO08T6B8exsBs618DoskFIChEsHqBNX9ITInxE"
ENDPOINT = "https://api.linear.app/graphql"

DOD_REFERENCE = "\n---\n\n> **Before opening a PR:** Complete all checks in the [Definition of Done](./ICE_ENGINEERING.md#3-definition-of-done) (`ICE_ENGINEERING.md` section 3). Self-review first, then request Peter's approval."

def gql(query, variables=None, retries=3):
    payload = {"query": query}
    if variables:
        payload["variables"] = variables
    for attempt in range(1, retries + 1):
        try:
            req = urllib.request.Request(ENDPOINT,
                data=json.dumps(payload).encode(),
                headers={"Content-Type": "application/json", "Authorization": API_KEY})
            resp = urllib.request.urlopen(req, timeout=30)
            data = json.loads(resp.read())
            if "errors" in data and data["errors"]:
                raise Exception(data["errors"][0]["message"])
            return data["data"]
        except Exception as e:
            if attempt == retries:
                raise
            wait = attempt * 2
            print(f"   ⏳ Retry {attempt}/{retries} in {wait}s... ({e})")
            time.sleep(wait)

print("🧊 ICE — Replacing inline DoD with reference link")
print("═" * 47 + "\n")

# 1. Fetch all issues
data = gql('{ issues(filter: { team: { key: { eq: "ICE" } } }, first: 50) { nodes { id identifier title description } } }')
issues = sorted(
    [i for i in data["issues"]["nodes"] if int(i["identifier"].replace("ICE-", "")) >= 29],
    key=lambda x: int(x["identifier"].replace("ICE-", ""))
)
print(f"Found {len(issues)} issues\n")

updated = 0
skipped = 0

for issue in issues:
    desc = issue["description"] or ""
    ident = issue["identifier"]

    if "## Definition of Done" not in desc:
        print(f"   ⏭️  {ident} — no DoD found, skipping")
        skipped += 1
        continue

    # Find the DoD section
    dod_index = desc.index("## Definition of Done")

    # Find the --- separator before it (within 10 chars)
    before = desc[:dod_index]
    last_sep = before.rfind("---")

    if last_sep != -1 and (dod_index - last_sep) < 10:
        cut_point = last_sep
    else:
        cut_point = dod_index

    # Trim trailing newlines before cut point
    while cut_point > 0 and desc[cut_point - 1] == "\n":
        cut_point -= 1

    preserved = desc[:cut_point]

    # Safety checks
    if len(preserved) < 200:
        print(f"   ⚠️  {ident} — preserved too short ({len(preserved)} chars), SKIPPING")
        skipped += 1
        continue

    if "## Goal" not in preserved or "## PR" not in preserved:
        print(f"   ⚠️  {ident} — missing Goal or PR after trim, SKIPPING")
        skipped += 1
        continue

    new_desc = preserved + DOD_REFERENCE
    removed_lines = desc.count("\n") - new_desc.count("\n")

    # Update via GraphQL mutation
    mutation = """mutation($id: String!, $input: IssueUpdateInput!) {
        issueUpdate(id: $id, input: $input) { success }
    }"""
    gql(mutation, {"id": issue["id"], "input": {"description": new_desc}})

    updated += 1
    short = (issue["title"].split("] ")[1] if "] " in issue["title"] else issue["title"])[:55]
    print(f"   ✅ {ident} — {short} (-{removed_lines} lines)")
    time.sleep(0.5)

print(f"\n{'═' * 47}")
print(f"🎉 Done! {updated} issues updated, {skipped} skipped")
print(f"\n📄 DoD now lives in: ICE_ENGINEERING.md (section 3)")
print(f"📎 Each task has a one-line reference pointing there")

File: scripts/cleanup-dod.mjs
#!/usr/bin/env node
/**
 * ICE — Replace inline DoD with a short reference to ICE_ENGINEERING.md
 * Carefully removes only the "## Definition of Done" section from each task,
 * preserving everything else (including P01's Tooling Setup section).
 */

const API_KEY = "lin_api_FEzO08T6B8exsBs618DoskFIChEsHqBNX9ITInxE";
const ENDPOINT = "https://api.linear.app/graphql";

async function gql(query, variables = {}, retries = 3) {
  for (let attempt = 1; attempt <= retries; attempt++) {
    try {
      const res = await fetch(ENDPOINT, {
        method: "POST",
        headers: { "Content-Type": "application/json", Authorization: API_KEY },
        body: JSON.stringify({ query, variables }),
        signal: AbortSignal.timeout(30000),
      });
      const json = await res.json();
      if (json.errors) throw new Error(json.errors[0].message);
      return json.data;
    } catch (err) {
      if (attempt === retries) throw err;
      console.log(`   ⏳ Retry ${attempt}/${retries}... (${err.message})`);
      await new Promise(r => setTimeout(r, attempt * 2000));
    }
  }
}

const delay = (ms) => new Promise(r => setTimeout(r, ms));

// The short reference that replaces the full DoD
const DOD_REFERENCE = [
  '',
  '---',
  '',
  '> **Before opening a PR:** Complete all checks in the [Definition of Done](./ICE_ENGINEERING.md#3-definition-of-done) (`ICE_ENGINEERING.md` section 3). Self-review first, then request Peter\'s approval.',
].join('\n');

async function main() {
  console.log('🧊 ICE — Replacing inline DoD with reference link');
  console.log('═══════════════════════════════════════════════\n');

  // 1. Fetch all issues
  const data = await gql(`{
    issues(filter: { team: { key: { eq: "ICE" } } }, first: 50) {
      nodes { id identifier title description }
    }
  }`);

  const issues = data.issues.nodes
    .filter(i => parseInt(i.identifier.replace('ICE-', '')) >= 29)
    .sort((a, b) => parseInt(a.identifier.replace('ICE-', '')) - parseInt(b.identifier.replace('ICE-', '')));

  console.log(`Found ${issues.length} issues\n`);

  let updated = 0;
  let skipped = 0;

  for (const issue of issues) {
    const desc = issue.description || '';

    // Check if DoD is present
    if (!desc.includes('## Definition of Done')) {
      console.log(`   ⏭️  ${issue.identifier} — no DoD found, skipping`);
      skipped++;
      continue;
    }

    // Find the DoD section and the --- separator before it
    // Pattern: look for "\n---\n\n## Definition of Done" and remove everything from there
    const dodMarker = '## Definition of Done';
    const dodIndex = desc.indexOf(dodMarker);

    if (dodIndex === -1) {
      console.log(`   ⏭️  ${issue.identifier} — DoD marker not found`);
      skipped++;
      continue;
    }

    // Find the --- separator that precedes the DoD heading
    // Search backwards from dodIndex for "---"
    const beforeDod = desc.substring(0, dodIndex);
    const lastSepIndex = beforeDod.lastIndexOf('---');

    let cutPoint;
    if (lastSepIndex !== -1 && dodIndex - lastSepIndex < 10) {
      // The --- is close to the DoD heading (within ~10 chars, accounting for newlines)
      // Cut from the --- separator
      cutPoint = lastSepIndex;
    } else {
      // No nearby separator, cut from the DoD heading itself
      cutPoint = dodIndex;
    }

    // Also remove any trailing newlines before the cut point
    let cleanedEnd = cutPoint;
    while (cleanedEnd > 0 && desc[cleanedEnd - 1] === '\n') {
      cleanedEnd--;
    }

    const preservedContent = desc.substring(0, cleanedEnd);
    const newDesc = preservedContent + DOD_REFERENCE;

    // Safety check: make sure we didn't accidentally nuke the description
    const originalLines = desc.split('\n').length;
    const newLines = newDesc.split('\n').length;
    const removedLines = originalLines - newLines;

    if (preservedContent.length < 200) {
      console.log(`   ⚠️  ${issue.identifier} — preserved content too short (${preservedContent.length} chars), SKIPPING for safety`);
      skipped++;
      continue;
    }

    // Verify key sections are still present
    const hasGoal = newDesc.includes('## Goal');
    const hasPR = newDesc.includes('## PR');
    if (!hasGoal || !hasPR) {
      console.log(`   ⚠️  ${issue.identifier} — missing Goal or PR section after trim, SKIPPING for safety`);
      skipped++;
      continue;
    }

    // Update
    await gql(`
      mutation($id: String!, $input: IssueUpdateInput!) {
        issueUpdate(id: $id, input: $input) { success }
      }
    `, { id: issue.id, input: { description: newDesc } });

    updated++;
    const shortTitle = (issue.title.split('] ')[1] || issue.title).slice(0, 55);
    console.log(`   ✅ ${issue.identifier} — ${shortTitle} (removed ${removedLines} lines)`);
    await delay(400);
  }

  console.log('\n═══════════════════════════════════════════════');
  console.log(`🎉 Done! ${updated} issues updated, ${skipped} skipped`);
  console.log(`\n📄 DoD now lives in: ICE_ENGINEERING.md (section 3)`);
  console.log(`📎 Each task has a one-line reference pointing there`);
}

main().catch(err => {
  console.error('\n💥 Fatal error:', err.message);
  process.exit(1);
});
